### ISSUE TITLE
FieldCodegen unexpectedly annotated

### ISSUE THREAD

**Issue by adetaylor** [google/autocxx#260]
Reproduction:

* Follow instructions in the `README` here: https://github.com/google/autocxx/compare/ortools-experiment
* Build like this: `touch build.rs && RUST_BACKTRACE=1 RUST_LOG=info cargo build 2>&1`

Outcome:

```
 thread 'main' panicked at 'Type was unexpectedly annotated', /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:3439:13
  stack backtrace:
     0: std::panicking::begin_panic
               at /rustc/3f5aee2d5241139d808f4fdece0026603489afd1/library/std/src/panicking.rs:519:12
     1: autocxx_bindgen::codegen::RustTy::unwrap_ts
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:3439:13
     2: <autocxx_bindgen::ir::comp::FieldData as autocxx_bindgen::codegen::FieldCodegen>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:1220:22
     3: <autocxx_bindgen::ir::comp::Field as autocxx_bindgen::codegen::FieldCodegen>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:1163:17
     4: <autocxx_bindgen::ir::comp::CompInfo as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:1759:17
     5: <autocxx_bindgen::ir::ty::Type as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:769:39
     6: <autocxx_bindgen::ir::item::Item as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:477:17
     7: <autocxx_bindgen::ir::comp::CompInfo as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:1995:13
     8: <autocxx_bindgen::ir::ty::Type as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:769:39
     9: <autocxx_bindgen::ir::item::Item as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:477:17
    10: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:499:21
    11: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:547:13
    12: autocxx_bindgen::codegen::CodegenResult::inner
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:318:9
    13: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:534:27
    14: <autocxx_bindgen::ir::item::Item as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:468:17
    15: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:499:21
    16: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:547:13
    17: autocxx_bindgen::codegen::CodegenResult::inner
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:318:9
    18: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:534:27
    19: <autocxx_bindgen::ir::item::Item as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:468:17
    20: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:499:21
    21: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:547:13
    22: autocxx_bindgen::codegen::CodegenResult::inner
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:318:9
    23: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:534:27
    24: <autocxx_bindgen::ir::item::Item as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:468:17
    25: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:499:21
    26: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:547:13
    27: autocxx_bindgen::codegen::CodegenResult::inner
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:318:9
    28: <autocxx_bindgen::ir::module::Module as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:534:27
    29: <autocxx_bindgen::ir::item::Item as autocxx_bindgen::codegen::CodeGenerator>::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:468:17
    30: autocxx_bindgen::codegen::codegen::{{closure}}
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:4209:9
    31: autocxx_bindgen::ir::context::BindgenContext::gen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/ir/context.rs:1159:19
    32: autocxx_bindgen::codegen::codegen
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/codegen/mod.rs:4183:5
    33: autocxx_bindgen::Bindings::generate
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/lib.rs:2207:32
    34: autocxx_bindgen::Builder::generate
               at /usr/local/google/home/adetaylor/.cargo/registry/src/github.com-1ecc6299db9ec823/autocxx-bindgen-0.57.1/src/lib.rs:1380:9
    35: autocxx_engine::IncludeCppEngine::generate
               at /usr/local/google/home/adetaylor/autocxx/engine/src/lib.rs:386:24
    36: autocxx_engine::parse_file::ParsedFile::resolve_all
               at /usr/local/google/home/adetaylor/autocxx/engine/src/parse_file.rs:125:13
    37: autocxx_engine::builder::build_to_custom_directory
               at /usr/local/google/home/adetaylor/autocxx/engine/src/builder.rs:136:5
    38: autocxx_engine::builder::build
               at /usr/local/google/home/adetaylor/autocxx/engine/src/builder.rs:75:5
    39: autocxx_build::build
               at /usr/local/google/home/adetaylor/autocxx/gen/build/src/lib.rs:33:5
    40: build_script_build::main
               at ./build.rs:23:17
    41: core::ops::function::FnOnce::call_once
               at /rustc/3f5aee2d5241139d808f4fdece0026603489afd1/library/core/src/ops/function.rs:227:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

```

**Comment by adetaylor** [comment]
(as reported in #222)

**Comment by adetaylor** [comment]
OK. or-tools, as it brings in absl and protobufs, turns out to be a fairly ambitious codebase which yields a lot of interesting findings :) I've found the following issues:

Need fixing before any part of this or-tools codebase will work: #260, #262, #263, #268, #269 and a workaround for #266
bindgen bug which I think we can work around: #264
Required before certain specific APIs will work: #259, #261, #265, #266
Nice to have: #267

I suspect we'll also need #102 which is blocked on a cxx change.

As I said in #222, I expected some rough edges. I'm only spending a small fraction of my time on autocxx so it'll take a while for me to get fixes in place for these.

**Comment by adetaylor** [comment]
@lovasoa here's the state of play:
I believe these items are fixed/worked around: #259, #260, #262, #263, #268 and #269, plus we workaround #266 and #264 as necessary.

#102 also landed.

_In theory_ it should now be possible to use `autocxx` with or-tools and have it produce _some_ callable APIs. However, some remaining caveats:
* I've not tried this yet. #271 says we should have a more complex demo example, and or-tools might be a good case, but I haven't done that.
* I've noticed that or-tools requires `std::vector<T*>` for some of its APIs which isn't yet supported either by cxx or autocxx - that's #330 and won't be especially imminent. You could write manual bindings for these APIs but #254 says that it could be easier than it currently is.
* #261, #265 and #266 mean some other APIs won't yet automatically translate.

So... I'm not going to comment here again but I thought I should give you an update. Thanks for finding a helpful test case to dig into these things! If _I_ do anything more with or-tools it will be on #271, but given the need for `std::vector<T*>` it might not be the codebase I choose.

**Comment by lovasoa** [comment]
@adetaylor That's awesome, thank you very much for all this work! I'm going to test that as soon as it's available on crates.io !
